src/psyclone/psyGen.py

```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
         return Literal(node.items[0], parent=parent)
 
 
+class Symbol(object):
+    '''
+    Symbol item for the Symbol Table. It contains information about: the name
+    of the symbol, its datatype, the number of dimensions and the symbol kind
+    (e.g. local, external, read_arg, write_arg, rw_arg).
+
+    :param name: Name of the symbol.
+    :type name: string
+    :param datatype: Data type of the symbol.
+    :type datatype: string
+    :param dimensions: Dimensions of the symbol (0 represents an scalar
+                       symbol).
+    :type dimensions: list of integers
+    :param kind: Information about the variable declaration attributes.
+    :type kind: string
+    '''
+    def __init__(self, name, datatype=None, dimensions=0, kind=None):
+        self._name = name
+        self._datatype = datatype
+        self._dimensions = dimensions
+        self._kind = kind
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def datatype(self):
+        return self._datatype
+
+    @property
+    def dimensions(self):
+        return self._dimensions
+
+    @property
+    def kind(self):
+        return self._kind
+
+    def __str__(self):
+        return (self.name + "<" + self.datatype + "," + str(self.dimensions) +
+                "," + self.kind + ">")
+
+
+class SymbolTable(object):
+    '''
+    Encapsulates the symbols table and provides methods to declare new symbols
+    and look up existing symbols. It is implemented as a monolithic scope
+    symbol table.
+    '''
+    def __init__(self):
+        self._symbols = {}
+
+    def declare(self, name, datatype, dimensions, kind):
+        '''
+        Declare a new symbol in the symbols table.
+
+        :param name: Name of the symbol.
+        :type name: string
+        :param datatype: Data type of the symbol.
+        :type datatype: string
+        :param dimensions: Dimensions of the symbol (0 represents an scalar
+                           symbol).
+        :type dimensions: list of integers
+        :param kind: Information about the variable declaration attributes.
+        :type kind: string
+        '''
+        if name in self._symbols:
+            raise InternalError("Multiple definition of symbol: '{0}'."
+                                "".format(name))
+        else:
+            self._symbols[name] = Symbol(name, datatype, dimensions, kind)
+
+    def lookup(self, name):
+        '''
+        Look up a symbol in the symbols table.
+
+        :param name: Name of the symbol
+        :type name: string
+        :raises KeyError: If the given name is not in the Symbol Table.
+        '''
+        return self._symbols[name]
+
+    def view(self):
+        '''
+        Print a representation of this Symbol Table to stdout.
+        '''
+        print("Symbol Table:")
```

Comment:
Could you do `print(str(self))` instead and just make use of the `__str__` method?

Reply:
I did it because I think it will often be used in conjunction with the schedule view, and from the 2 scripts below the former looks nicer:
```
kernel_code = kernel.get_kernel_schedule()
kernel_code.symbol_table.view()
kernel_code.view()
```
```
kernel_code = kernel.get_kernel_schedule()
print(str(kernel_code.symbol_table))
kernel_code.view()
```
But it is true than functionally they are the same. I am happy with either option.

Reply:
I think what I meant was that the view() method could just print the results of the str() method:

    def view(self):
        print(str(self))

Does that work?